### ISSUE TITLE
Add "nix processes" command that is equivalent to "guix processes"

### ISSUE THREAD

**Issue by matthewbauer** [NixOS/nix#3751]
Guix has a feature that is missing in Nix:

https://guix.gnu.org/manual/en/html_node/Invoking-guix-processes.html

`guix processes` is used to list derivations that are running by the daemon. It's useful to system administrators who want to know what is using up so much of their resources.

**Comment by Kloenk** [comment]
There is as an adhoc solution [nix-top](https://github.com/samueldr/nix-top). But yeah, a real nix subcommand could be usefully (or at least protocol definition to query this info from the daemon)

**Comment by zimbatm** [comment]
Related to that, it would be great if nix would report back *what* is blocking a build when it is locked or waiting on a runner to become available.

**Comment by Kloenk** [comment]
One simple idea which would not be too complicated in implementing is writing the current build step into `/nix/var/nix/userpool/${build-user}`.

Is that a good idea?

**Comment by Atemu** [comment]
I like the idea of it being discoverable via files. Though perhaps they should live in `/run` instead, this kind of information doesn't need to be held permanently.

**Comment by Kloenk** [comment]
> I like the idea of it being discoverable via files. Though perhaps they should live in `/run` instead, this kind of information doesn't need to be held permanently.

Of course. My idea behind that was that it would make it easier to see if the look on the user is already held without calling flock. As the locks live under /nix/var/nix/userpool

**Comment by matthewbauer** [comment]
I made a quick proof of concept here:

https://github.com/NixOS/nix/pull/3800

I think handling build steps would be useful too. Right now you can check /proc for child processes, but it only works on systems with a procfs.

**Comment by Ericson2314** [comment]
I assume we limit to trusted users?

**Comment by matthewbauer** [comment]
userpool is only readable by root (or whoever owns NIX_STORE_DIR) so any new information we expose would have that restriction. I don’t think it’s worth letting trusted-users see it. I don’t think there’s anything stopping them from doing `pstree $(pgrep nix-daemon)` though.

**Comment by Atemu** [comment]
Still important to me.

**Comment by Kloenk** [comment]
Still important to me

**Comment by Atemu** [comment]
Still important to me.
